common/src/ray_tracing.rs

```diff
@@ -0,0 +1,105 @@
+use crate::math;
+
+/// Performs ray tracing against the z=0 plane in node coordinates. This plane is one-sided, so the ray must be going in the -z
+/// direction to detect an intersection.
+///
+/// The `start_node_transform` parameter determines which coordinate system the `ray` parameter and any resulting intersection
+/// normals are given in. Specifically, the `start_node_transform` matrix converts this coordinate system to node coordinates.
+///
+/// The `tanh_distance` is the hyperbolic tangent of the distance along the ray to check for intersections.
+pub fn trace_ray(
```

Comment:
This function *demands* a decent suite of unit tests.

Reply:
The logic that definitely needs unit tests in this function gets moved to `sphere_collider.rs` later on. My plan for now is to create a unit test in this commit and move/adapt it in the future commit. It's possible we'll still want this method to be tested directly in the final result, but I'll hold off on that for now.

Reply:
Playtesting is an extremely inefficient way to test. We should cover at least basic sanity checks, and ideally interesting corner cases as well.

Reply:
I'm happy to create a few basic sanity checks. However, the fact that we want test suites on functions that act on a populated `DualGraph` suggests that a future enhancement outside the scope of this PR would be the ability to select a "test world" for worldgen, and this can be the same world that is used for these kinds of tests.

For now, I can create a basic test world with only two or three solid blocks, although this will still be a fair amount of setup code.

Reply:
Specifying geometry to test against is certainly necessary. We should work towards factoring out common code for that as test cases grow.

Reply:
I managed to add a test case for this method which tests a fair amount of its functionality. To save my own time, my plan is to wait until you've reviewed some of my tests before I add some for corner cases. That way, I can get a better sense of your preferred coding style when it comes to tests for this project.

Reply:
Based on your review, it doesn't seem like there's anything fundamentally wrong with how I set up these tests, so I'll add some edge case tests here before I request another review.

Reply:
I found a bug while testing an edge case, so you were definitely right that this needed testing. When setting `ray_node_distance`, I used the wrong formula for "distance to the origin". For a normalized hypervector, it should either by `.xyz().magnitude().asinh()` or `.w.cosh()`, but I used `.xyz().magnitude().acosh()`.

I tried to test for this manually, but I guess my aim wasn't good enough. Thanks for insisting on unit tests!